python/pyarrow/array.pxi

```diff
@@ -1699,6 +1711,81 @@ cdef class Array(_PandasConvertible):
                                                      c_type))
         return pyarrow_wrap_array(c_array)
 
+    def __arrow_c_array__(self, requested_schema=None):
+        """
+        Get a pair of PyCapsules containing a C ArrowArray representation of the object.
+
+        Parameters
+        ----------
+        requested_schema : PyCapsule | None
+            A PyCapsule containing a C ArrowSchema representation of a requested
+            schema. PyArrow will attempt to cast the array to this data type.
+            If None, the array will be returned as-is, with a type matching the
+            one returned by :meth:`__arrow_c_schema__()`.
+
+        Returns
+        -------
+        Tuple[PyCapsule, PyCapsule]
+            A pair of PyCapsules containing a C ArrowSchema and ArrowArray,
+            respectively.
+        """
+        cdef:
+            ArrowArray* c_array = <ArrowArray*> malloc(sizeof(ArrowArray))
+            ArrowSchema* c_schema = <ArrowSchema*> malloc(sizeof(ArrowSchema))
+
+        if requested_schema is not None:
+            target_type = DataType._import_from_c_capsule(requested_schema)
+
+            if target_type != self.type:
+                try:
+                    casted_array = _pc().cast(self, target_type, safe=True)
+                    inner_array = pyarrow_unwrap_array(casted_array)
```

Comment:
Is Cython able to infer the type for `inner_array`? I see no `cdef` for it.

Reply:
TBH I have no idea how to tell what Cython is inferring. Doesn't quite have the IDE support C++ or Rust has. I've added a `cdef` for it just in case 🤷